skimage/transform/_seam_carving.pyx

```diff
@@ -0,0 +1,217 @@
+        cumulative_img[0, c] = energy_img[0, c, 0]
+
+    for r in range(1, rows):
+        for c in range(cols):
+            min_cost = DBL_MAX
+            for offset in range(-1, 2):
+
+                c_idx = c + offset
+                if (c_idx > cols - 1) or (c_idx < 0):
+                    continue
+
+                if cumulative_img[r-1, c_idx] < min_cost:
+                    min_cost = cumulative_img[r-1, c_idx]
+                    track_img[r, c] = offset
+
+            cumulative_img[r, c] = min_cost + energy_img[r, c, 0]
+
+cdef cnp.uint8_t _mark_seam(cnp.int8_t[:, ::1] track_img,
+                            Py_ssize_t start_index,
+                            cnp.uint8_t[:, ::1] seam_map):
+
+    """ Re-trace the optimal seam from a given column in the last row.
+
+    This function tries to re-track an optimal seam from `start_index` and
+    tries to mark it in `seam_map`. If this seam intersects with any existing
+    seam in `seam_map` the function returns `0` without marking anything. Else
+    it marks the seam in `seam_map` and returns `1`.
+
+    track_img : (M, N) ndarray
+       The array of relative column indices as updated by `_preprocess_image`.
+    start_index : int
+       The column number of the bottom most row from where to start re-tracing
+       the seam.
+    seam_map : (M, N) ndarray
+        The array used to mark seams. If a pixel is marked as as seam it is set
+        to `1`, else `0`.
+
+    Returns
+    -------
+    out : int
+        `1` if seam was marked, `0` is seam inersects and was not marked.
+    """
+    cdef Py_ssize_t rows = track_img.shape[0]
+    cdef Py_ssize_t[::1] current_seam_indices = np.zeros(rows, dtype=np.int)
+    cdef Py_ssize_t row, col
+    cdef cnp.int8_t offset
+    cdef Py_ssize_t seams
+
+    current_seam_indices[rows - 1] = start_index
+    for row in range(rows - 2, -1, -1):
+        col = current_seam_indices[row+1]
+        offset = track_img[row, col]
+        col = col + offset
+        current_seam_indices[row] = col
+
+        if seam_map[row, col]:
+            return 0
+
+    for row in range(rows):
+        col = current_seam_indices[row]
+        seam_map[row, col] = 1
+
+    return 1
+
+cdef _remove_seam(cnp.double_t[:, :, ::1] img,
+                  cnp.uint8_t[:, ::1] seam_map, Py_ssize_t cols):
+    """ Removes marked seams from an image.
+
+    Parameters
+    ----------
+    img : (M, N, P) ndarray
+        Input image whose vertical seams are to be removed.
+    seam_map : (M, N) ndarray
+        Array with seams to be removed marked by non-zero entries.
+    cols : int
+        The number of colums to process.
+    """
+    cdef Py_ssize_t rows = img.shape[0]
+    cdef Py_ssize_t channels = img.shape[2]
+    cdef Py_ssize_t r, c, ch, shift
+
+    for r in range(rows):
+        shift = 0
+        for c in range(cols):
+            shift += seam_map[r, c]
+            for ch in range(channels):
+                img[r, c, ch] = img[r, c + shift, ch]
```

Comment:
Looks like `c + shift` could be precalculated after `shift` and referenced from one loop level prior to the innermost.